x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartDataFrameAnalyticsAction.java

```diff
@@ -572,7 +582,10 @@ protected AllocatedPersistentTask createTask(
             String id = params.getId();
 
             List<String> unavailableIndices =
-                verifyIndicesPrimaryShardsAreActive(clusterState, resolver, AnomalyDetectorsIndex.configIndexName());
+                verifyIndicesPrimaryShardsAreActive(clusterState,
+                    resolver,
+                    AnomalyDetectorsIndex.configIndexName(),
+                    MlStatsIndex.indexPattern());
```

Comment:
I feel like I'm missing something obvious here. The `.ml-stats` index is created by [ DataFrameAnalyticsManager.execute](https://github.com/elastic/elasticsearch/blob/dd331935b31f83024633a6ec84da5be946b4e7d1/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/DataFrameAnalyticsManager.java#L114) called from `TaskExecutor.nodeOperation`.   But `TaskExecutor.getAssignment` which runs first is expecting the index to exist and as far as I can tell it hasn't been created yet. Could that be because of the leniency in `verifyIndicesPrimaryShardsAreActive` 


`resolver.concreteIndexNames(clusterState, IndicesOptions.lenientExpandOpen(), indexNames);`

Reply:
Yes, arguably we could prevent assignment if the required indices exist but are not usable at the time an assignment decision is made.  But the changes currently in this PR will not fix #53007.

> we could prevent assignment if the required indices exist but are not usable at the time an assignment decision is made

A side effect of this would be that the start datafeed API would fail in some cases where if it had waited a fraction of a second it would have succeeded.  I am not sure that's a good idea.  It certainly deserves a discussion.

Reply:
For Transforms we get the stats directly from the task if it is running else the index is searched. Could we not do the same here? In which case `.ml-stats` would not be touched until the DFA has finished.

I think for anomaly detection we created most of the indices on node start (apart from the results index) which is one way of solving the problem but is heavy handed and still required checks that the index was active.

Reply:
There is a scenario y'all are not thinking about. 

What if the previously created index is just not available due to the nodes being down?

Its fine being lenient in `verifyIndicesPrimaryShardsAreActive` if the index does not exist at all. This is because we will create it. 

But, if it already exists and its primary shard is not assigned, that is a bigger issue. One warranting the task not being assigned to a node.

> For Transforms we get the stats directly from the task if it is running else the index is searched. Could we not do the same here?

We could. I can look into that. 

Regardless of waiting for the indices to be created before returning from `_start`, I think the changes in `verifyIndicesPrimaryShardsAreActive` are necessary and good.

Reply:
> What if the previously created index is just not available due to the nodes being down?

This is what I was referring to in https://github.com/elastic/elasticsearch/pull/53029#discussion_r386943414

So it solves one problem but creates another:

> A side effect of this would be that the start datafeed API would fail in some cases where if it had waited a fraction of a second it would have succeeded. I am not sure that's a good idea. It certainly deserves a discussion.

I think if we do this we should do it consistently for anomaly detection and data frame analytics, so that both or neither assign tasks when the necessary indices exist but are not available.

@dimitris-athanasiou I think you originally added some of these checks.  Do you have any extra insights?

Reply:
> A side effect of this would be that the start datafeed API would fail in some cases where if it had waited a fraction of a second it would have succeeded.

It won't fail. It just won't get assigned a node. Which means that it will attempt to get assigned again in a little bit

Reply:
> For Transforms we get the stats directly from the task if it is running else the index is searched.

We already get the stats for running tasks from in-memory.

> @dimitris-athanasiou I think you originally added some of these checks. Do you have any extra insights?

I think not checking whether the state and stats indices have their primaries active there is an omission. But it is not the cause of this error.

@droberts195, @benwtrent and myself discussed this further and we decided we will need to add debugging in order to understand what exactly is going on as it is currently not perfectly clear.

Reply:
> changes in verifyIndicesPrimaryShardsAreActive are necessary and good.

Yes I agree this is what we do for AD and we should check the same indices here.